src/district/server/smart_contracts.cljs

```diff
@@ -115,132 +107,134 @@
-                 (try
-                   (when (and (string? result)
-                              (map? (last args)))
-                     (loop [gas-used (handle-contract-call method result)]
-                       (when-not gas-used
-                         (recur (handle-contract-call method result (:print-gas-usage? @smart-contracts))))))
-                   (web3-eth/stop-watching! @filter-id)
-                   (callback nil result)
-                   (catch js/Error err
-                     (callback err)))))))))
-
-(def contract-call-deasynced (deasync contract-call*))
-
-(defn contract-call [& args]
-  (if (:auto-mining? @smart-contracts)
-    (apply contract-call* args)
-    (apply contract-call-deasynced args)))
+;; TODO : add alts for definite timeout (or # of retries)
+(defn- wait-for-tx-receipt*
+  "callback is a nodejs style callback i.e. (fn [error data] ...)"
+  [tx-hash callback]
+  (web3-eth/get-transaction-receipt @web3 tx-hash (fn [error receipt]
+                                                    (if error
+                                                      (callback error nil)
+                                                      (go
+                                                        (if receipt
+                                                          (callback nil receipt)
+                                                          (do
+                                                            ;; try again in 1K millis
+                                                            (<! (timeout 1000))
+                                                            (wait-for-tx-receipt* tx-hash callback))))))))
+
+(defn wait-for-tx-receipt
+  "blocks until transaction `tx-hash` gets sent to the network."
+  [tx-hash]
+  (js/Promise. (fn [resolve reject]
+                 (wait-for-tx-receipt* tx-hash (fn [error tx-receipt]
+                                                 (if error
+                                                   (reject error)
+                                                   (resolve tx-receipt)))))))
+
+(defn contract-call
+  "# arguments:
+   ## `contract` parameter can be one of:
+   * keyword :some-contract
+   * tuple of keyword and address [:some-contract 0x1234...]
+   * instance SomeContract
+   ## `method` is a :camel_case keyword corresponding to the smart-contract function
+   ## `args` is a vector of arguments for the `method`
+   ## `opts` is a map of options passed as message data
+   # returns:
+   function returns a Promise"
+  ([contract method args {:keys [:from :gas] :as opts}]
+   (let [opts (merge (when-not from
+                       {:from (first (web3-eth/accounts @web3))})
+                     (when-not gas
+                       {:gas 4000000})
+                     opts)]
+     (js/Promise. (fn [resolve reject]
+                    (apply web3-eth/contract-call (instance-from-arg contract) method
+                           (merge args
```

Comment:
I think this is better like
```clojure
(into args [opts
            (fn [err data]
              (if err
                (reject err)
                (resolve data)))])
```
since merge is not  well defined for non maps parameters

Reply:
done.